File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include "iostream"
#include "stack"
#include "string"

using namespace std;

int main() {
    while (true) {
        string str;
        getline(cin, str);
        if (str == ".") {
            break;
        }
        stack<char> st;
        bool check = true;
        for (int i = 0; i < str.length(); i++) {
            char c = str[i];
            if (c == '(' || c == '[') {
                st.push(c);
            } else if (c == ')') {
                if (!st.empty() && st.top() == '(') {
                    st.pop();
                } else {
                    check = false;
                    break;
                }
            } else if (c == ']') {
                if (!st.empty() && st.top() == '[') {
                    st.pop();
                } else {
                    check = false;
                    break;
                }
            }
        }
        if (check && st.empty()) {
            cout << "yes" << "\n";
        } else {
            cout << "no" << "\n";
        }
    }
    return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            String input = br.readLine();
            if (input.equals(".")) break;
            boolean ans = true;
            Stack<Character> st = new Stack<>();
            for (int s = 0; s < input.length(); s++) {
                Character ch = input.charAt(s);
                if (ch == '(' || ch == '[') {
                    st.add(ch);
                } else if (ch == ')') {
                    if (st.isEmpty() || st.peek()  == '[') {
                        ans = false;
                        break;
                    } else {
                        st.pop();
                    }
                } else if (ch == ']') {
                    if (st.isEmpty() || st.peek() == '(') {
                        ans = false;
                        break;
                    } else {
                        st.pop();
                    }
                }
            }
            if (!st.isEmpty()) ans = false;
            
            if (ans) System.out.println("yes");
            else System.out.println("no");
        }
    }
}
